scripts/ci-entrypoint.sh

```diff
@@ -146,6 +148,19 @@ get_cidrs() {
     fi
 }
 
+# get_cloud_provider determines if the Cluster is using an intree or external cloud-provider from the KubeadmConfigSpec.
+# any retryable operation in this function must return a non-zero exit code on failure so that we can
+# retry it using a `until get_cloud_provider; do sleep 5; done` pattern;
+# and any statement must be idempotent so that subsequent retry attempts can make forward progress.
+get_cloud_provider() {
+    # TODO: make this retryable.
```

Comment:
A couple things:

1) We should probably be initiating a more specific get (the one below will work because as of now the kind mgmt cluster is created and only hosts one cluster, so there's only one resource of type `kubeadmcontrolplane`)
2) It seems that `kubectl` does not incorporate the terminal outcome of the jsonpath lookup into returned exit code.

For example, against a local mgmt cluster of mine:

```bash
Francisbook:cluster-api-provider-azure jackfrancis$ k get kubeadmcontrolplane/capz-ajfz73-control-plane -n default -o=jsonpath='{.spec.kubeadmConfigSpec.clusterConfiguration.controllerManager.extraArgs.foo}'
Francisbook:cluster-api-provider-azure jackfrancis$ echo $?
0
```

So, I think we can do this:

If the exit code is non-zero, then the resource doesn't exist (or the GET failed). For example:

```bash
Francisbook:cluster-api-provider-azure jackfrancis$ k get kubeadmcontrolplane/does-not-exist -n default -o=jsonpath='{.spec.kubeadmConfigSpec.clusterConfiguration.controllerManager.extraArgs.cloud-provider}'
Error from server (NotFound): kubeadmcontrolplanes.controlplane.cluster.x-k8s.io "does-not-exist" not found
Francisbook:cluster-api-provider-azure jackfrancis$ echo $?
1
Francisbook:cluster-api-provider-azure jackfrancis$ export KUBECONFIG=notexist
Francisbook:cluster-api-provider-azure jackfrancis$ k get kubeadmcontrolplane/does-not-exist -n default -o=jsonpath='{.spec.kubeadmConfigSpec.clusterConfiguration.controllerManager.extraArgs.cloud-provider}'
W0227 15:57:20.123773   72284 loader.go:222] Config not found: notexist
The connection to the server localhost:8080 was refused - did you specify the right host or port?
Unable to get version information from cluster
Francisbook:cluster-api-provider-azure jackfrancis$ echo $?
1
```

If the result is an empty string, we may assume we have a non-existent, or misconfigured property:

```bash
Francisbook:cluster-api-provider-azure jackfrancis$ [[ "" == $(k get kubeadmcontrolplane/capz-ajfz73-control-plane -n default -o=jsonpath='{.spec.kubeadmConfigSpec.clusterConfiguration.controllerManager.extraArgs.foo}') ]]
Francisbook:cluster-api-provider-azure jackfrancis$ echo $?
0
```

Because there are two known-working values here, we can probably just set up a ternary condition to do the various things based on known values ("external", "azure", or empty string, to anticipate the property going away), and then an OR condition to trigger a retry. Something like this demonstrates the nominal outcome when we get one of our known values:

```bash
Francisbook:cluster-api-provider-azure jackfrancis$ RESULT=$(k get kubeadmcontrolplane/capz-ajfz73-control-plane -n default -o=jsonpath='{.spec.kubeadmConfigSpec.clusterConfiguration.controllerManager.extraArgs.cloud-provider}') && echo $RESULT || echo "should retry"
external
```

And then this shows that the same logic should detect our retry outcome:

```bash
Francisbook:cluster-api-provider-azure jackfrancis$ RESULT=$(k get kubeadmcontrolplane/does-not-exist-should-retry -n default -o=jsonpath='{.spec.kubeadmConfigSpec.clusterConfiguration.controllerManager.extraArgs.cloud-provider}') && echo $RESULT || echo "should retry"
Error from server (NotFound): kubeadmcontrolplanes.controlplane.cluster.x-k8s.io "does-not-exist-should-retry" not found
should retry
```

Reply:
I think that should do it, PTAL